conformance-tests/src/main.rs

```diff
@@ -0,0 +1,257 @@
+            stderr,
+            io_mode: IoMode::Http(port),
+        };
+        let start = std::time::Instant::now();
+        loop {
+            match std::net::TcpStream::connect(format!("127.0.0.1:{port}")) {
+                Ok(_) => {
+                    log::debug!("Spin shim started on port {port}.");
+                    return Ok(spin);
+                }
+                Err(e) => {
+                    let stderr = spin.stderr.output_as_str().unwrap_or("<non-utf8>");
+                    log::trace!("Checking that the shim server started returned an error: {e}");
+                    log::trace!("Current spin stderr = '{stderr}'");
+                }
+            }
+            if let Some(status) = spin.try_wait()? {
+                anyhow::bail!(
+                    "Shim exited early with status code {:?}\n{}{}",
+                    status.code(),
+                    spin.stdout.output_as_str().unwrap_or("<non-utf8>"),
+                    spin.stderr.output_as_str().unwrap_or("<non-utf8>")
+                );
+            }
+
+            if start.elapsed() > std::time::Duration::from_secs(2 * 60) {
+                break;
+            }
+            std::thread::sleep(std::time::Duration::from_millis(50));
+        }
+        anyhow::bail!(
+            "`ctr run` did not start server or error after two minutes. stderr:\n\t{}",
+            spin.stderr.output_as_str().unwrap_or("<non-utf8>")
+        )
+    }
+
+    /// Make an HTTP request against Spin
+    ///
+    /// Will fail if Spin has already exited or if the io mode is not HTTP
+    pub fn make_http_request(&mut self, request: Request<'_, String>) -> anyhow::Result<Response> {
+        let IoMode::Http(port) = self.io_mode;
+        if let Some(status) = self.try_wait()? {
+            anyhow::bail!(
+                "make_http_request - shim exited early with status code {:?}",
+                status.code()
+            );
+        }
+        log::debug!("Connecting to HTTP server on port {port}...");
+        let response = request.send("localhost", port)?;
+        log::debug!("Awaiting response from server");
+        if let Some(status) = self.try_wait()? {
+            anyhow::bail!("Spin exited early with status code {:?}", status.code());
+        }
+        Ok(response)
+    }
+
+    pub fn stderr(&mut self) -> &str {
+        self.stderr.output_as_str().unwrap_or("<non-utf8>")
+    }
+
+    fn try_wait(&mut self) -> std::io::Result<Option<std::process::ExitStatus>> {
+        self.process.try_wait()
+    }
+}
+
+impl Drop for SpinShim {
+    fn drop(&mut self) {
+        kill_process(&mut self.process);
+    }
+}
+
+impl Runtime for SpinShim {
+    fn error(&mut self) -> anyhow::Result<()> {
+        if self.try_wait()?.is_some() {
+            anyhow::bail!("Containerd shim spin exited early: {}", self.stderr());
+        }
+
+        Ok(())
+    }
+}
+
+/// How this instance is communicating with the outside world
+pub enum IoMode {
```

Comment:
Will the test harness support non-HTTP workloads as well?

Reply:
Yes. It should eventually support command and maybe SQS if we can mock a provider. Spin's supports redis, [for example](https://github.com/fermyon/spin/blob/444adbaac3271a358da560c077d2c25c756b3056/tests/testing-framework/src/runtimes/spin_cli.rs#L252).